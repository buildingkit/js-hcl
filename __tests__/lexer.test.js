const lexer = require('../lib/lexer')

describe('lexer', () => {
  it('should contain block name token', () => {
    const data = `service {}`;
    const tokens = lexer(data).run();

    expect(tokens[0].value).toBe('service');
    expect(tokens[2].value).toBe('{');
    expect(tokens[3].value).toBe('}');
  });

  it('should cointain dynamic block name token', () => {
    const data = `service "group" {}`;
    const tokens = lexer(data).run();

    expect(tokens[0].value).toBe('service');
    expect(tokens[2].value).toBe('group');
    expect(tokens[4].value).toBe('{');
    expect(tokens[5].value).toBe('}');
  });

  it('should contain block name with equal token', () => {
    const data = `service = {}`;
    const tokens = lexer(data).run();

    expect(tokens[0].value).toBe('service');
    expect(tokens[2].value).toBe('=');
    expect(tokens[4].value).toBe('{');
    expect(tokens[5].value).toBe('}');
  })

  it('should scan docker input', () => {
    const data = `group "default" {
      targets = ["db", "webapp-dev"]
    }
    target "webapp-dev" {
      dockerfile = "Dockerfile.webapp"
      tags = ["docker.io/username/webapp"]
    }
    target "webapp-release" {
      inherits = ["webapp-dev"]
      platforms = ["linux/amd64", "linux/arm64"]
    }
    target "db" {
      dockerfile = "Dockerfile.db"
      tags = ["docker.io/username/db"]
    }`
    const tokens = lexer(data).run();
    expect(tokens).toMatchSnapshot();
  })
})
