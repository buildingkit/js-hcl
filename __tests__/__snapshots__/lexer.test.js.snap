// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lexer should scan docker input 1`] = `
Array [
  Object {
    "column": 1,
    "id": "block",
    "length": 5,
    "line": 1,
    "value": "group",
  },
  Object {
    "column": 6,
    "id": "whitespace",
    "length": 1,
    "line": 1,
    "value": " ",
  },
  Object {
    "column": 7,
    "id": "string",
    "length": 9,
    "line": 1,
    "value": "default",
  },
  Object {
    "column": 16,
    "id": "whitespace",
    "length": 1,
    "line": 1,
    "value": " ",
  },
  Object {
    "column": 17,
    "id": "left_bracket",
    "length": 1,
    "line": 1,
    "value": "{",
  },
  Object {
    "column": 18,
    "id": "newline",
    "length": 1,
    "line": 1,
    "value": "
",
  },
  Object {
    "column": 1,
    "id": "whitespace",
    "length": 1,
    "line": 2,
    "value": " ",
  },
  Object {
    "column": 2,
    "id": "whitespace",
    "length": 1,
    "line": 2,
    "value": " ",
  },
  Object {
    "column": 3,
    "id": "whitespace",
    "length": 1,
    "line": 2,
    "value": " ",
  },
  Object {
    "column": 4,
    "id": "whitespace",
    "length": 1,
    "line": 2,
    "value": " ",
  },
  Object {
    "column": 5,
    "id": "whitespace",
    "length": 1,
    "line": 2,
    "value": " ",
  },
  Object {
    "column": 6,
    "id": "whitespace",
    "length": 1,
    "line": 2,
    "value": " ",
  },
  Object {
    "column": 7,
    "id": "name",
    "length": 7,
    "line": 2,
    "value": "targets",
  },
  Object {
    "column": 14,
    "id": "whitespace",
    "length": 1,
    "line": 2,
    "value": " ",
  },
  Object {
    "column": 15,
    "id": "equal",
    "length": 1,
    "line": 2,
    "value": "=",
  },
  Object {
    "column": 16,
    "id": "whitespace",
    "length": 1,
    "line": 2,
    "value": " ",
  },
  Object {
    "column": 17,
    "id": "array_open",
    "length": 1,
    "line": 2,
    "value": "[",
  },
  Object {
    "column": 18,
    "id": "string",
    "length": 4,
    "line": 2,
    "value": "db",
  },
  Object {
    "column": 22,
    "id": "comma",
    "length": 1,
    "line": 2,
    "value": ",",
  },
  Object {
    "column": 23,
    "id": "whitespace",
    "length": 1,
    "line": 2,
    "value": " ",
  },
  Object {
    "column": 24,
    "id": "string",
    "length": 12,
    "line": 2,
    "value": "webapp-dev",
  },
  Object {
    "column": 36,
    "id": "array_close",
    "length": 1,
    "line": 2,
    "value": "]",
  },
  Object {
    "column": 37,
    "id": "newline",
    "length": 1,
    "line": 2,
    "value": "
",
  },
  Object {
    "column": 1,
    "id": "whitespace",
    "length": 1,
    "line": 3,
    "value": " ",
  },
  Object {
    "column": 2,
    "id": "whitespace",
    "length": 1,
    "line": 3,
    "value": " ",
  },
  Object {
    "column": 3,
    "id": "whitespace",
    "length": 1,
    "line": 3,
    "value": " ",
  },
  Object {
    "column": 4,
    "id": "whitespace",
    "length": 1,
    "line": 3,
    "value": " ",
  },
  Object {
    "column": 5,
    "id": "right_bracket",
    "length": 1,
    "line": 3,
    "value": "}",
  },
  Object {
    "column": 6,
    "id": "newline",
    "length": 1,
    "line": 3,
    "value": "
",
  },
  Object {
    "column": 1,
    "id": "whitespace",
    "length": 1,
    "line": 4,
    "value": " ",
  },
  Object {
    "column": 2,
    "id": "whitespace",
    "length": 1,
    "line": 4,
    "value": " ",
  },
  Object {
    "column": 3,
    "id": "whitespace",
    "length": 1,
    "line": 4,
    "value": " ",
  },
  Object {
    "column": 4,
    "id": "whitespace",
    "length": 1,
    "line": 4,
    "value": " ",
  },
  Object {
    "column": 5,
    "id": "block",
    "length": 6,
    "line": 4,
    "value": "target",
  },
  Object {
    "column": 11,
    "id": "whitespace",
    "length": 1,
    "line": 4,
    "value": " ",
  },
  Object {
    "column": 12,
    "id": "string",
    "length": 12,
    "line": 4,
    "value": "webapp-dev",
  },
  Object {
    "column": 24,
    "id": "whitespace",
    "length": 1,
    "line": 4,
    "value": " ",
  },
  Object {
    "column": 25,
    "id": "left_bracket",
    "length": 1,
    "line": 4,
    "value": "{",
  },
  Object {
    "column": 26,
    "id": "newline",
    "length": 1,
    "line": 4,
    "value": "
",
  },
  Object {
    "column": 1,
    "id": "whitespace",
    "length": 1,
    "line": 5,
    "value": " ",
  },
  Object {
    "column": 2,
    "id": "whitespace",
    "length": 1,
    "line": 5,
    "value": " ",
  },
  Object {
    "column": 3,
    "id": "whitespace",
    "length": 1,
    "line": 5,
    "value": " ",
  },
  Object {
    "column": 4,
    "id": "whitespace",
    "length": 1,
    "line": 5,
    "value": " ",
  },
  Object {
    "column": 5,
    "id": "whitespace",
    "length": 1,
    "line": 5,
    "value": " ",
  },
  Object {
    "column": 6,
    "id": "whitespace",
    "length": 1,
    "line": 5,
    "value": " ",
  },
  Object {
    "column": 7,
    "id": "name",
    "length": 10,
    "line": 5,
    "value": "dockerfile",
  },
  Object {
    "column": 17,
    "id": "whitespace",
    "length": 1,
    "line": 5,
    "value": " ",
  },
  Object {
    "column": 18,
    "id": "equal",
    "length": 1,
    "line": 5,
    "value": "=",
  },
  Object {
    "column": 19,
    "id": "whitespace",
    "length": 1,
    "line": 5,
    "value": " ",
  },
  Object {
    "column": 20,
    "id": "string",
    "length": 19,
    "line": 5,
    "value": "Dockerfile.webapp",
  },
  Object {
    "column": 39,
    "id": "newline",
    "length": 1,
    "line": 5,
    "value": "
",
  },
  Object {
    "column": 1,
    "id": "whitespace",
    "length": 1,
    "line": 6,
    "value": " ",
  },
  Object {
    "column": 2,
    "id": "whitespace",
    "length": 1,
    "line": 6,
    "value": " ",
  },
  Object {
    "column": 3,
    "id": "whitespace",
    "length": 1,
    "line": 6,
    "value": " ",
  },
  Object {
    "column": 4,
    "id": "whitespace",
    "length": 1,
    "line": 6,
    "value": " ",
  },
  Object {
    "column": 5,
    "id": "whitespace",
    "length": 1,
    "line": 6,
    "value": " ",
  },
  Object {
    "column": 6,
    "id": "whitespace",
    "length": 1,
    "line": 6,
    "value": " ",
  },
  Object {
    "column": 7,
    "id": "name",
    "length": 4,
    "line": 6,
    "value": "tags",
  },
  Object {
    "column": 11,
    "id": "whitespace",
    "length": 1,
    "line": 6,
    "value": " ",
  },
  Object {
    "column": 12,
    "id": "equal",
    "length": 1,
    "line": 6,
    "value": "=",
  },
  Object {
    "column": 13,
    "id": "whitespace",
    "length": 1,
    "line": 6,
    "value": " ",
  },
  Object {
    "column": 14,
    "id": "array_open",
    "length": 1,
    "line": 6,
    "value": "[",
  },
  Object {
    "column": 15,
    "id": "string",
    "length": 27,
    "line": 6,
    "value": "docker.io/username/webapp",
  },
  Object {
    "column": 42,
    "id": "array_close",
    "length": 1,
    "line": 6,
    "value": "]",
  },
  Object {
    "column": 43,
    "id": "newline",
    "length": 1,
    "line": 6,
    "value": "
",
  },
  Object {
    "column": 1,
    "id": "whitespace",
    "length": 1,
    "line": 7,
    "value": " ",
  },
  Object {
    "column": 2,
    "id": "whitespace",
    "length": 1,
    "line": 7,
    "value": " ",
  },
  Object {
    "column": 3,
    "id": "whitespace",
    "length": 1,
    "line": 7,
    "value": " ",
  },
  Object {
    "column": 4,
    "id": "whitespace",
    "length": 1,
    "line": 7,
    "value": " ",
  },
  Object {
    "column": 5,
    "id": "right_bracket",
    "length": 1,
    "line": 7,
    "value": "}",
  },
  Object {
    "column": 6,
    "id": "newline",
    "length": 1,
    "line": 7,
    "value": "
",
  },
  Object {
    "column": 1,
    "id": "whitespace",
    "length": 1,
    "line": 8,
    "value": " ",
  },
  Object {
    "column": 2,
    "id": "whitespace",
    "length": 1,
    "line": 8,
    "value": " ",
  },
  Object {
    "column": 3,
    "id": "whitespace",
    "length": 1,
    "line": 8,
    "value": " ",
  },
  Object {
    "column": 4,
    "id": "whitespace",
    "length": 1,
    "line": 8,
    "value": " ",
  },
  Object {
    "column": 5,
    "id": "block",
    "length": 6,
    "line": 8,
    "value": "target",
  },
  Object {
    "column": 11,
    "id": "whitespace",
    "length": 1,
    "line": 8,
    "value": " ",
  },
  Object {
    "column": 12,
    "id": "string",
    "length": 16,
    "line": 8,
    "value": "webapp-release",
  },
  Object {
    "column": 28,
    "id": "whitespace",
    "length": 1,
    "line": 8,
    "value": " ",
  },
  Object {
    "column": 29,
    "id": "left_bracket",
    "length": 1,
    "line": 8,
    "value": "{",
  },
  Object {
    "column": 30,
    "id": "newline",
    "length": 1,
    "line": 8,
    "value": "
",
  },
  Object {
    "column": 1,
    "id": "whitespace",
    "length": 1,
    "line": 9,
    "value": " ",
  },
  Object {
    "column": 2,
    "id": "whitespace",
    "length": 1,
    "line": 9,
    "value": " ",
  },
  Object {
    "column": 3,
    "id": "whitespace",
    "length": 1,
    "line": 9,
    "value": " ",
  },
  Object {
    "column": 4,
    "id": "whitespace",
    "length": 1,
    "line": 9,
    "value": " ",
  },
  Object {
    "column": 5,
    "id": "whitespace",
    "length": 1,
    "line": 9,
    "value": " ",
  },
  Object {
    "column": 6,
    "id": "whitespace",
    "length": 1,
    "line": 9,
    "value": " ",
  },
  Object {
    "column": 7,
    "id": "name",
    "length": 8,
    "line": 9,
    "value": "inherits",
  },
  Object {
    "column": 15,
    "id": "whitespace",
    "length": 1,
    "line": 9,
    "value": " ",
  },
  Object {
    "column": 16,
    "id": "equal",
    "length": 1,
    "line": 9,
    "value": "=",
  },
  Object {
    "column": 17,
    "id": "whitespace",
    "length": 1,
    "line": 9,
    "value": " ",
  },
  Object {
    "column": 18,
    "id": "array_open",
    "length": 1,
    "line": 9,
    "value": "[",
  },
  Object {
    "column": 19,
    "id": "string",
    "length": 12,
    "line": 9,
    "value": "webapp-dev",
  },
  Object {
    "column": 31,
    "id": "array_close",
    "length": 1,
    "line": 9,
    "value": "]",
  },
  Object {
    "column": 32,
    "id": "newline",
    "length": 1,
    "line": 9,
    "value": "
",
  },
  Object {
    "column": 1,
    "id": "whitespace",
    "length": 1,
    "line": 10,
    "value": " ",
  },
  Object {
    "column": 2,
    "id": "whitespace",
    "length": 1,
    "line": 10,
    "value": " ",
  },
  Object {
    "column": 3,
    "id": "whitespace",
    "length": 1,
    "line": 10,
    "value": " ",
  },
  Object {
    "column": 4,
    "id": "whitespace",
    "length": 1,
    "line": 10,
    "value": " ",
  },
  Object {
    "column": 5,
    "id": "whitespace",
    "length": 1,
    "line": 10,
    "value": " ",
  },
  Object {
    "column": 6,
    "id": "whitespace",
    "length": 1,
    "line": 10,
    "value": " ",
  },
  Object {
    "column": 7,
    "id": "name",
    "length": 9,
    "line": 10,
    "value": "platforms",
  },
  Object {
    "column": 16,
    "id": "whitespace",
    "length": 1,
    "line": 10,
    "value": " ",
  },
  Object {
    "column": 17,
    "id": "equal",
    "length": 1,
    "line": 10,
    "value": "=",
  },
  Object {
    "column": 18,
    "id": "whitespace",
    "length": 1,
    "line": 10,
    "value": " ",
  },
  Object {
    "column": 19,
    "id": "array_open",
    "length": 1,
    "line": 10,
    "value": "[",
  },
  Object {
    "column": 20,
    "id": "string",
    "length": 13,
    "line": 10,
    "value": "linux/amd64",
  },
  Object {
    "column": 33,
    "id": "comma",
    "length": 1,
    "line": 10,
    "value": ",",
  },
  Object {
    "column": 34,
    "id": "whitespace",
    "length": 1,
    "line": 10,
    "value": " ",
  },
  Object {
    "column": 35,
    "id": "string",
    "length": 13,
    "line": 10,
    "value": "linux/arm64",
  },
  Object {
    "column": 48,
    "id": "array_close",
    "length": 1,
    "line": 10,
    "value": "]",
  },
  Object {
    "column": 49,
    "id": "newline",
    "length": 1,
    "line": 10,
    "value": "
",
  },
  Object {
    "column": 1,
    "id": "whitespace",
    "length": 1,
    "line": 11,
    "value": " ",
  },
  Object {
    "column": 2,
    "id": "whitespace",
    "length": 1,
    "line": 11,
    "value": " ",
  },
  Object {
    "column": 3,
    "id": "whitespace",
    "length": 1,
    "line": 11,
    "value": " ",
  },
  Object {
    "column": 4,
    "id": "whitespace",
    "length": 1,
    "line": 11,
    "value": " ",
  },
  Object {
    "column": 5,
    "id": "right_bracket",
    "length": 1,
    "line": 11,
    "value": "}",
  },
  Object {
    "column": 6,
    "id": "newline",
    "length": 1,
    "line": 11,
    "value": "
",
  },
  Object {
    "column": 1,
    "id": "whitespace",
    "length": 1,
    "line": 12,
    "value": " ",
  },
  Object {
    "column": 2,
    "id": "whitespace",
    "length": 1,
    "line": 12,
    "value": " ",
  },
  Object {
    "column": 3,
    "id": "whitespace",
    "length": 1,
    "line": 12,
    "value": " ",
  },
  Object {
    "column": 4,
    "id": "whitespace",
    "length": 1,
    "line": 12,
    "value": " ",
  },
  Object {
    "column": 5,
    "id": "block",
    "length": 6,
    "line": 12,
    "value": "target",
  },
  Object {
    "column": 11,
    "id": "whitespace",
    "length": 1,
    "line": 12,
    "value": " ",
  },
  Object {
    "column": 12,
    "id": "string",
    "length": 4,
    "line": 12,
    "value": "db",
  },
  Object {
    "column": 16,
    "id": "whitespace",
    "length": 1,
    "line": 12,
    "value": " ",
  },
  Object {
    "column": 17,
    "id": "left_bracket",
    "length": 1,
    "line": 12,
    "value": "{",
  },
  Object {
    "column": 18,
    "id": "newline",
    "length": 1,
    "line": 12,
    "value": "
",
  },
  Object {
    "column": 1,
    "id": "whitespace",
    "length": 1,
    "line": 13,
    "value": " ",
  },
  Object {
    "column": 2,
    "id": "whitespace",
    "length": 1,
    "line": 13,
    "value": " ",
  },
  Object {
    "column": 3,
    "id": "whitespace",
    "length": 1,
    "line": 13,
    "value": " ",
  },
  Object {
    "column": 4,
    "id": "whitespace",
    "length": 1,
    "line": 13,
    "value": " ",
  },
  Object {
    "column": 5,
    "id": "whitespace",
    "length": 1,
    "line": 13,
    "value": " ",
  },
  Object {
    "column": 6,
    "id": "whitespace",
    "length": 1,
    "line": 13,
    "value": " ",
  },
  Object {
    "column": 7,
    "id": "name",
    "length": 10,
    "line": 13,
    "value": "dockerfile",
  },
  Object {
    "column": 17,
    "id": "whitespace",
    "length": 1,
    "line": 13,
    "value": " ",
  },
  Object {
    "column": 18,
    "id": "equal",
    "length": 1,
    "line": 13,
    "value": "=",
  },
  Object {
    "column": 19,
    "id": "whitespace",
    "length": 1,
    "line": 13,
    "value": " ",
  },
  Object {
    "column": 20,
    "id": "string",
    "length": 15,
    "line": 13,
    "value": "Dockerfile.db",
  },
  Object {
    "column": 35,
    "id": "newline",
    "length": 1,
    "line": 13,
    "value": "
",
  },
  Object {
    "column": 1,
    "id": "whitespace",
    "length": 1,
    "line": 14,
    "value": " ",
  },
  Object {
    "column": 2,
    "id": "whitespace",
    "length": 1,
    "line": 14,
    "value": " ",
  },
  Object {
    "column": 3,
    "id": "whitespace",
    "length": 1,
    "line": 14,
    "value": " ",
  },
  Object {
    "column": 4,
    "id": "whitespace",
    "length": 1,
    "line": 14,
    "value": " ",
  },
  Object {
    "column": 5,
    "id": "whitespace",
    "length": 1,
    "line": 14,
    "value": " ",
  },
  Object {
    "column": 6,
    "id": "whitespace",
    "length": 1,
    "line": 14,
    "value": " ",
  },
  Object {
    "column": 7,
    "id": "name",
    "length": 4,
    "line": 14,
    "value": "tags",
  },
  Object {
    "column": 11,
    "id": "whitespace",
    "length": 1,
    "line": 14,
    "value": " ",
  },
  Object {
    "column": 12,
    "id": "equal",
    "length": 1,
    "line": 14,
    "value": "=",
  },
  Object {
    "column": 13,
    "id": "whitespace",
    "length": 1,
    "line": 14,
    "value": " ",
  },
  Object {
    "column": 14,
    "id": "array_open",
    "length": 1,
    "line": 14,
    "value": "[",
  },
  Object {
    "column": 15,
    "id": "string",
    "length": 23,
    "line": 14,
    "value": "docker.io/username/db",
  },
  Object {
    "column": 38,
    "id": "array_close",
    "length": 1,
    "line": 14,
    "value": "]",
  },
  Object {
    "column": 39,
    "id": "newline",
    "length": 1,
    "line": 14,
    "value": "
",
  },
  Object {
    "column": 1,
    "id": "whitespace",
    "length": 1,
    "line": 15,
    "value": " ",
  },
  Object {
    "column": 2,
    "id": "whitespace",
    "length": 1,
    "line": 15,
    "value": " ",
  },
  Object {
    "column": 3,
    "id": "whitespace",
    "length": 1,
    "line": 15,
    "value": " ",
  },
  Object {
    "column": 4,
    "id": "whitespace",
    "length": 1,
    "line": 15,
    "value": " ",
  },
  Object {
    "column": 5,
    "id": "right_bracket",
    "length": 1,
    "line": 15,
    "value": "}",
  },
]
`;
